File: services/auth-service/internal/service/oauth_config_service.go
package service

import (
	"time"

	"github.com/beegy-labs/my-girok/services/auth-service/internal/model"
	"github.com/beegy-labs/my-girok/services/auth-service/internal/repository"
)

type OAuthConfigService struct {
	oauthConfigRepo *repository.OAuthConfigRepository
}

func NewOAuthConfigService(oauthConfigRepo *repository.OAuthConfigRepository) *OAuthConfigService {
	return &OAuthConfigService{
		oauthConfigRepo: oauthConfigRepo,
	}
}

// GetAllProviders returns all OAuth provider configurations
func (s *OAuthConfigService) GetAllProviders() ([]model.OAuthProviderConfig, error) {
	return s.oauthConfigRepo.FindAll()
}

// GetEnabledProviders returns only enabled OAuth providers
func (s *OAuthConfigService) GetEnabledProviders() ([]model.OAuthProviderConfig, error) {
	return s.oauthConfigRepo.FindEnabled()
}

// GetProvider returns a specific OAuth provider configuration
func (s *OAuthConfigService) GetProvider(provider model.AuthProvider) (*model.OAuthProviderConfig, error) {
	return s.oauthConfigRepo.FindByProvider(provider)
}

// ToggleProvider enables or disables an OAuth provider
func (s *OAuthConfigService) ToggleProvider(provider model.AuthProvider, enabled bool, updatedBy string) (*model.OAuthProviderConfig, error) {
	config, err := s.oauthConfigRepo.FindByProvider(provider)
	if err != nil {
		// Create new config if not exists
		config = &model.OAuthProviderConfig{
			Provider:    provider,
			Enabled:     enabled,
			DisplayName: getProviderDisplayName(provider),
			Description: getProviderDescription(provider),
			UpdatedAt:   time.Now(),
			UpdatedBy:   updatedBy,
		}
		if err := s.oauthConfigRepo.Create(config); err != nil {
			return nil, err
		}
		return config, nil
	}

	config.Enabled = enabled
	config.UpdatedAt = time.Now()
	config.UpdatedBy = updatedBy

	if err := s.oauthConfigRepo.Update(config); err != nil {
		return nil, err
	}

	return config, nil
}

// IsProviderEnabled checks if a specific OAuth provider is enabled
func (s *OAuthConfigService) IsProviderEnabled(provider model.AuthProvider) (bool, error) {
	return s.oauthConfigRepo.IsProviderEnabled(provider)
}

// UpdateProviderConfig updates OAuth provider configuration
func (s *OAuthConfigService) UpdateProviderConfig(
	provider model.AuthProvider,
	displayName, description, callbackURL, updatedBy string,
) (*model.OAuthProviderConfig, error) {
	config, err := s.oauthConfigRepo.FindByProvider(provider)
	if err != nil {
		return nil, err
	}

	if displayName != "" {
		config.DisplayName = displayName
	}
	if description != "" {
		config.Description = description
	}
	if callbackURL != "" {
		config.CallbackURL = callbackURL
	}
	config.UpdatedAt = time.Now()
	config.UpdatedBy = updatedBy

	if err := s.oauthConfigRepo.Update(config); err != nil {
		return nil, err
	}

	return config, nil
}

func getProviderDisplayName(provider model.AuthProvider) string {
	names := map[model.AuthProvider]string{
		model.ProviderGoogle: "Google",
		model.ProviderKakao:  "Kakao",
		model.ProviderNaver:  "Naver",
		model.ProviderApple:  "Apple",
	}
	if name, ok := names[provider]; ok {
		return name
	}
	return string(provider)
}

func getProviderDescription(provider model.AuthProvider) string {
	descriptions := map[model.AuthProvider]string{
		model.ProviderGoogle: "Sign in with Google account",
		model.ProviderKakao:  "Sign in with Kakao account",
		model.ProviderNaver:  "Sign in with Naver account",
		model.ProviderApple:  "Sign in with Apple ID",
	}
	if desc, ok := descriptions[provider]; ok {
		return desc
	}
	return ""
}

File: services/auth-service/src/oauth-config/oauth-config.service.ts
import {
  Injectable,
  NotFoundException,
  ForbiddenException,
  BadRequestException,
  Logger,
} from '@nestjs/common';
import { PrismaService } from '../database/prisma.service';
import { AuthProvider } from '@my-girok/types';
import { CryptoService } from '../common/crypto/crypto.service';
import { OAuthProviderResponseDto, EnabledProviderDto, EnabledProvidersResponseDto } from './dto';

@Injectable()
export class OAuthConfigService {
  private readonly logger = new Logger(OAuthConfigService.name);

  // Allowed callback URL domains
  private readonly allowedCallbackDomains = [
    'localhost',
    'girok.dev',
    'auth.girok.dev',
    'auth-bff.girok.dev',
  ];

  constructor(
    private prisma: PrismaService,
    private crypto: CryptoService,
  ) {}

  /**
   * Check if OAuth provider is enabled
   * LOCAL provider is always enabled
   * If config doesn't exist, default to enabled
   */
  async isProviderEnabled(provider: AuthProvider): Promise<boolean> {
    // LOCAL provider is always enabled
    if (provider === AuthProvider.LOCAL) {
      return true;
    }

    const config = await this.prisma.oAuthProviderConfig.findUnique({
      where: { provider },
    });

    // If no config exists, default to enabled
    if (!config) {
      return true;
    }

    return config.enabled;
  }

  /**
   * Get all OAuth provider configurations
   */
  async getAllProviders() {
    return this.prisma.oAuthProviderConfig.findMany({
      orderBy: { provider: 'asc' },
    });
  }

  /**
   * Toggle OAuth provider on/off
   * Only MASTER admins can toggle providers
   * LOCAL provider cannot be disabled
   */
  async toggleProvider(provider: AuthProvider, enabled: boolean, adminUserId: string) {
    // Prevent disabling LOCAL provider
    if (provider === AuthProvider.LOCAL && !enabled) {
      throw new ForbiddenException('LOCAL provider cannot be disabled');
    }

    // Check if config exists
    const existingConfig = await this.prisma.oAuthProviderConfig.findUnique({
      where: { provider },
    });

    if (existingConfig) {
      // Update existing config
      return this.prisma.oAuthProviderConfig.update({
        where: { provider },
        data: {
          enabled,
          updatedBy: adminUserId,
        },
      });
    } else {
      // Create new config
      return this.prisma.oAuthProviderConfig.upsert({
        where: { provider },
        create: {
          provider,
          enabled,
          displayName: this.getDisplayName(provider),
          description: `Login with ${this.getDisplayName(provider)}`,
          updatedBy: adminUserId,
        },
        update: {
          enabled,
          updatedBy: adminUserId,
        },
      });
    }
  }

  /**
   * Get specific provider configuration
   */
  async getProviderConfig(provider: AuthProvider) {
    const config = await this.prisma.oAuthProviderConfig.findUnique({
      where: { provider },
    });

    if (!config) {
      throw new NotFoundException(`OAuth provider ${provider} configuration not found`);
    }

    return config;
  }

  /**
   * Helper: Get display name for provider
   */
  private getDisplayName(provider: AuthProvider): string {
    const displayNames: Record<AuthProvider, string> = {
      [AuthProvider.LOCAL]: 'Local',
      [AuthProvider.GOOGLE]: 'Google',
      [AuthProvider.KAKAO]: 'Kakao',
      [AuthProvider.NAVER]: 'Naver',
      [AuthProvider.APPLE]: 'Apple',
    };
    return displayNames[provider];
  }

  /**
   * Update OAuth provider credentials
   * Encrypts client secret before storing
   */
  async updateProviderCredentials(
    provider: AuthProvider,
    data: {
      clientId?: string;
      clientSecret?: string;
      callbackUrl?: string;
    },
    adminUserId: string,
  ) {
    // Validate callback URL if provided
    if (data.callbackUrl) {
      this.validateCallbackUrl(data.callbackUrl);
    }

    // Encrypt client secret if provided
    const encryptedData: any = {};
    if (data.clientId !== undefined) {
      encryptedData.clientId = data.clientId;
    }
    if (data.clientSecret !== undefined) {
      encryptedData.clientSecret = this.crypto.encrypt(data.clientSecret);
    }
    if (data.callbackUrl !== undefined) {
      encryptedData.callbackUrl = data.callbackUrl;
    }

    // Check if config exists
    const existingConfig = await this.prisma.oAuthProviderConfig.findUnique({
      where: { provider },
    });

    if (existingConfig) {
      // Update existing config
      const updated = await this.prisma.oAuthProviderConfig.update({
        where: { provider },
        data: {
          ...encryptedData,
          updatedBy: adminUserId,
        },
      });

      this.logger.log(`OAuth credentials updated for provider ${provider} by admin ${adminUserId}`);

      return this.maskProviderConfig(updated);
    } else {
      // Create new config
      const created = await this.prisma.oAuthProviderConfig.create({
        data: {
          provider,
          ...encryptedData,
          displayName: this.getDisplayName(provider),
          description: `Login with ${this.getDisplayName(provider)}`,
          updatedBy: adminUserId,
        },
      });

      this.logger.log(`OAuth credentials created for provider ${provider} by admin ${adminUserId}`);

      return this.maskProviderConfig(created);
    }
  }

  /**
   * Get decrypted credentials for OAuth strategies
   * Used internally by strategies to initialize OAuth flow
   */
  async getDecryptedCredentials(provider: AuthProvider): Promise<{
    clientId: string | null;
    clientSecret: string | null;
    callbackUrl: string | null;
  }> {
    const config = await this.prisma.oAuthProviderConfig.findUnique({
      where: { provider },
    });

    if (!config) {
      return {
        clientId: null,
        clientSecret: null,
        callbackUrl: null,
      };
    }

    // Decrypt client secret if it exists
    let decryptedSecret: string | null = null;
    if (config.clientSecret) {
      try {
        decryptedSecret = this.crypto.decrypt(config.clientSecret);
      } catch (error) {
        this.logger.error(`Failed to decrypt secret for provider ${provider}: ${error}`);
        // Return null if decryption fails
        decryptedSecret = null;
      }
    }

    return {
      clientId: config.clientId,
      clientSecret: decryptedSecret,
      callbackUrl: config.callbackUrl,
    };
  }

  /**
   * Get enabled OAuth providers (public endpoint)
   * Returns only enabled providers without credentials
   */
  async getEnabledProviders(): Promise<EnabledProvidersResponseDto> {
    const allConfigs = await this.prisma.oAuthProviderConfig.findMany({
      where: {
        enabled: true,
        // Exclude LOCAL provider from OAuth list
        provider: {
          not: AuthProvider.LOCAL,
        },
      },
      orderBy: { provider: 'asc' },
    });

    const providers: EnabledProviderDto[] = allConfigs.map((config) => ({
      provider: config.provider as AuthProvider,
      displayName: config.displayName,
      description: config.description || undefined,
    }));

    return { providers };
  }

  /**
   * Get all providers with masked secrets (for admin UI)
   */
  async getAllProvidersWithMasking(): Promise<OAuthProviderResponseDto[]> {
    const allConfigs = await this.prisma.oAuthProviderConfig.findMany({
      orderBy: { provider: 'asc' },
    });

    return allConfigs.map((config) => this.maskProviderConfig(config));
  }

  /**
   * Mask provider config for safe display
   */
  private maskProviderConfig(config: any): OAuthProviderResponseDto {
    return {
      provider: config.provider as AuthProvider,
      enabled: config.enabled,
      clientId: config.clientId || undefined,
      clientSecretMasked: config.clientSecret ? this.maskSecret(config.clientSecret) : undefined,
      callbackUrl: config.callbackUrl || undefined,
      displayName: config.displayName,
      description: config.description || undefined,
      updatedAt: config.updatedAt,
      updatedBy: config.updatedBy || undefined,
    };
  }

  /**
   * Mask secret for display (show last 4 characters)
   */
  private maskSecret(secret: string): string {
    if (!secret || secret.length <= 4) {
      return '********';
    }
    const last4 = secret.slice(-4);
    return `********${last4}`;
  }

  /**
   * Validate callback URL against allowed domains
   */
  private validateCallbackUrl(url: string): void {
    try {
      const parsedUrl = new URL(url);
      const hostname = parsedUrl.hostname;

      // Check if hostname is in allowed domains or is subdomain of allowed domain
      const isAllowed = this.allowedCallbackDomains.some(
        (domain) =>
          hostname === domain ||
          hostname.endsWith(`.${domain}`) ||
          (domain === 'localhost' && hostname.startsWith('localhost')),
      );

      if (!isAllowed) {
        throw new BadRequestException(
          `Callback URL domain '${hostname}' is not in allowed domains: ${this.allowedCallbackDomains.join(', ')}`,
        );
      }
    } catch (error) {
      if (error instanceof BadRequestException) {
        throw error;
      }
      throw new BadRequestException(`Invalid callback URL format: ${url}`);
    }
  }
}
